msrest/polling/poller.py

```diff
@@ -0,0 +1,214 @@
+        self._polling_method = polling_method
+
+        if isinstance(deserialization_callback, type) and issubclass(deserialization_callback, Model):
+            deserialization_callback = deserialization_callback.deserialize
+
+        # Might raise a CloudError
+        self._polling_method.initialize(self._client, self._response, deserialization_callback)
+
+        # Prepare thread execution
+        self._thread = None
+        self._done = None
+        self._exception = None
+        if not self._polling_method.finished():
+            self._done = threading.Event()
+            self._thread = threading.Thread(
+                target=self._start,
+                name="LROPoller({})".format(uuid.uuid4()))
+            self._thread.daemon = True
+            self._thread.start()
+
+    def _start(self):
+        """Start the long running operation.
+        On completion, runs any callbacks.
+
+        :param callable update_cmd: The API reuqest to check the status of
+         the operation.
+        """
+        try:
+            self._polling_method.run()
+        except Exception as err:
+            self._exception = err
+
+        finally:
+            self._done.set()
+
+        callbacks, self._callbacks = self._callbacks, []
+        while callbacks:
+            for call in callbacks:
+                call(self._polling_method)
+            callbacks, self._callbacks = self._callbacks, []
+
+    def status(self):
+        """Returns the current status string.
+
+        :returns: The current status string
+        :rtype: str
+        """
+        return self._polling_method.status()
+
+    def result(self, timeout=None):
+        """Return the result of the long running operation, or
+        the result available after the specified timeout.
+
+        :returns: The deserialized resource of the long running operation,
+         if one is available.
+        :raises CloudError: Server problem with the query.
+        """
+        self.wait(timeout)
+        return self._polling_method.resource()
+
+    def wait(self, timeout=None):
+        """Wait on the long running operation for a specified length
+        of time.
+
+        :param int timeout: Perion of time to wait for the long running
+         operation to complete.
+        :raises CloudError: Server problem with the query.
+        """
+        if self._thread is None:
+            return
+        self._thread.join(timeout=timeout)
+        try:
+            raise self._exception
+        except TypeError: # Was None
+            pass
+
+    def done(self):
+        """Check status of the long running operation.
+
+        :returns: 'True' if the process has completed, else 'False'.
+        """
+        return self._thread is None or not self._thread.is_alive()
+
+    def add_done_callback(self, func):
+        """Add callback function to be run once the long running operation
+        has completed - regardless of the status of the operation.
+
```

Comment:
Maybe this comment could read: Add callback function to be run once the long running operation has completed . A callback can only be added before the operation has completed, and will be run regardless of whether the operation was successful.

Reply:
This should follow the asyncio behavior:
https://docs.python.org/3/library/asyncio-task.html#asyncio.Future.add_done_callback

If the future is already done when this is called, the callback is scheduled with call_soon().

Reply:
So I would change this code to execute the callback right now if it's done.

Reply:
Just to clarify - do you mean if _done is set? Or if the thread has finished? :)

Reply:
```python
        if self.done():
            func(self._polling_method)
        else:
            self._callbacks.append(func)

```